File: src/dymaptic.GeoBlazor.Core/Model/GoToOverride.cs
namespace dymaptic.GeoBlazor.Core.Model;

/// <summary>
///     The following properties define a goTo override function.
///     <a target="_blank" href="https://developers.arcgis.com/javascript/latest/api-reference/esri-widgets-support-widget.html#GoToOverride">ArcGIS Maps SDK for JavaScript</a>
/// </summary>
/// <param name="ViewId">
///     A reference Id to the MapView or SceneView where the navigation takes place.
/// </param>
/// <param name="Target">
///     The target location/viewpoint/extent to animate to.
/// </param>
/// <param name="Options">
///     Options defining the animation, duration, and easing of the navigation.
/// </param>
public record GoToOverrideParameters(Guid ViewId, GoToTarget Target, GoToOptions? Options);

/// <summary>
///     The target location/viewpoint to animate to in the goTo() method. A two or three-element array of numbers represents the [x,y,z] coordinates to center the view on. When using an object for the target, use the properties in this record.
/// </summary>
/// <param name="Target">
///     The target of the animation.
/// </param>
/// <param name="Center">
///     The MapView.Center to go to.
/// </param>
/// <param name="Scale">
///     The MapView.Scale to go to.
/// </param>
/// <param name="Zoom">
///     The MapView.Zoom to go to.
/// </param>
/// <param name="Heading">
///     The Camera.Heading to go to.
/// </param>
/// <param name="Tilt">
///     The Camera.Tilt to go to.
/// </param>
/// <param name="Position">
///     The Camera.Position to go to.
/// </param>
/// <param name="TargetGeometries">
///     The target as an array of Geometries.
/// </param>
/// <param name="TargetGraphics">
///     The target as an array of Graphics.
/// </param>
/// <param name="TargetGeometry">
///     The target as a single Geometry.
/// </param>
/// <param name="TargetGraphic">
///     The target as a single Graphic.
/// </param>
public record GoToTarget(
    object? Target,
    double[]? Center,
    double? Scale,
    double? Zoom,
    double? Heading,
    double? Tilt,
    Point? Position,
    Geometry[]? TargetGeometries,
    Graphic[]? TargetGraphics,
    Geometry? TargetGeometry,
    Graphic? TargetGraphic);

File: src/dymaptic.GeoBlazor.Core/Scripts/goToOverride.ts
import {hasValue, lookupGeoBlazorId} from './geoBlazorCore';
import {buildDotNetPoint} from "./point";
import Graphic from "@arcgis/core/Graphic";
import {buildDotNetGraphic} from "./graphic";
import Geometry from "@arcgis/core/geometry/Geometry";
import {buildDotNetGeometry} from "./geometry";

export function buildJsGoToOverride(dotNetObject: any, viewId: string | null) {
    return async (view: any, goToParameters: any) => {
        viewId ??= lookupGeoBlazorId(view);
        let parameters = buildDotNetGoToOverrideParameters(goToParameters, viewId!);
        await dotNetObject.dotNetComponentReference.invokeMethodAsync('OnJsGoToOverride', parameters);
    };
}

export async function buildDotNetGoToOverride(jsObject: any): Promise<any> {
    return null;
}

export function buildDotNetGoToOverrideParameters(parameters: any, viewId: string): any {
    let dnParams: any = {
        viewId: viewId,
        options: {
            animate: parameters.options.animate,
            duration: parameters.options.duration,
            easing: parameters.options.easing,
            pickClosestTarget: parameters.options.pickClosestTarget
        }
    }
    if (hasValue(parameters.target)) {
        let target: any = {
            target: parameters.target.target.toJSON(),
            center: parameters.target.center,
            scale: parameters.target.scale,
            zoom: parameters.target.zoom,
            heading: parameters.target.heading,
            tilt: parameters.target.tilt,
            position: buildDotNetPoint(parameters.target.position)
        };
        if (Array.isArray(parameters.target.target) && parameters.target.target.length > 0) {
            let firstObject = parameters.target.target[0];
            if (firstObject instanceof Graphic || firstObject.declaredClass.includes('graphic')) {
                target.targetGraphics = [];
                for (let g in parameters.target.target as Graphic[]) {
                    target.targetGraphics.push(buildDotNetGraphic(g as any, null, viewId));
                }
            } else if (firstObject instanceof Geometry || firstObject.declaredClass.includes('geometry')) {
                target.targetGeometries = [];
                for (let g in parameters.target.target as Geometry[]) {
                    target.targetGeometries.push(buildDotNetGeometry(g as any));
                }
            } else {
                target.targetCoordinates = parameters.target.target;
            }
        } else if (parameters.target.target instanceof Graphic || parameters.target.target.declaredClass.includes('graphic')) {
            target.targetGraphic = buildDotNetGraphic(parameters.target.target, null, viewId);
        } else if (parameters.target.target instanceof Geometry || parameters.target.target.declaredClass.includes('geometry')) {
            target.targetGeometry = buildDotNetGeometry(parameters.target.target);
        }
        dnParams.target = target;
    }

    return dnParams;
}
